### ISSUE TITLE
How much does DjangoQL improve performance?

### ISSUE THREAD

**Issue by caleb15** [ivelum/djangoql#41]
I'm presuming djangoql searches are much faster than regular django searches as you can select on the field you want rather than django's crazy "search everything" approach. But you do have any hard stats on how much performance is improved by? It would vary a lot based on amount of data and schema of course, but it would still be good to know.

**Comment by iceone** [comment]
Hi @caleb15!
Your thoughts were in the right direction. Although, this question isn't quite correct.
The DjangoQL is a little abstraction above plain Django lookups. So, a performance of your query depends on your DB schema (indexes, queryset optimizations), amount of data and your DB engine. And, the most important, it depends on the complexity of your DjangoQL query (and therefore the resulting SQL).

**Comment by caleb15** [comment]
Hi @iceone - I'm still a bit confused. I'm aware that query performance depends on schema, amount of data, DB engine, and complexity of the query. But DjangoQL searches on a specific field, thereby reducing the complexity (no need to search all fields), right?

**Comment by iceone** [comment]
As I mentioned, the question isn't correct. 
On a different input, the performance of DjangoQL and plain Django search may vary.
You can build a very specific query and it will very fast. You can build more common ILIKE query on a big text field in a large table and it will be very slow.
And in general, DjangoQL isn't about performance. It's about flexibility.

**Comment by caleb15** [comment]
Let me rephrase the question: Assuming the same search term, DB engine, schema, and data, how much faster would a DjangoQL query against a specific field be versus a plain django search?

Hopefully this is the right question, I'm fairly new to django so I may not be understanding the issue properly 😬

**Comment by iceone** [comment]
These queries will produce the same SQL. 
For example, DjangoQL: `title ~ "Foobar"` and plain django "Foobar" with the `search_fields = ['title']`

**Comment by caleb15** [comment]
I implemented djangoql for https://15five.15five.com

DjangoQL can _vastly_ speed up the search, depending mainly on how many records there are along with some other factors like you have mentioned.

```
Page      |  # Items   |  Performance
--------------------------------------------------------------------------
XXXX      |  ~1,000    |  no performance difference
YYYY      |  ~100,000  |  up to ~5x increase in search speed
ZZZZ      | ~1,000,000 |  searching can take so long it times out unless djangoql is used
```

Thanks for making DjangoQL @stebunovd and other contributors! I'm quite happy with the outcome.

**Comment by stebunovd** [comment]
Awesome! Thanks @caleb15, I'm glad it was helpful for your project
